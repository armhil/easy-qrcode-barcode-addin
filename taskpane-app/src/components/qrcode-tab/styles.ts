import { makeStyles } from '@fluentui/react-components';

export const useStyles = makeStyles({
  slider: {
    marginTop: '15px',
    marginBottom: '10px'
  },
  button: {
    marginTop: '15px'
  }
})