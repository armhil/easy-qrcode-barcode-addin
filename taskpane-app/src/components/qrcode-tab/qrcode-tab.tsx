import React, { useState, useRef } from 'react';
import { Button } from '@fluentui/react-components';
import { AddinUtils, LoggingUtils } from 'easy-addins-utils';
import QRCode from 'qrcode.react';
import { InputWrapper } from '../input';
import { Slider } from '@fluentui/react';
import { useStyles } from './styles';
// Qr code route
export function QrCodeTab() {
  React.useEffect(() => {
    LoggingUtils.Trace('qrbar-qrcode');
  }, []);

  const ref = useRef(null);
  const [text, setText] = useState('');
  const [size, setSize] = useState(125);
  const styles = useStyles();

  const insertImageFromCanvas = () => {
    const canvasValue = getCanvasURL();
    if (canvasValue) {
      AddinUtils.InsertImage(canvasValue, () => {});
    }
  };

  // Insert image to Word
  const insertImage = () => {
    if (!text || text.length === 0) {
      AddinUtils.GetText((t: string) => {
        setText(t);
        insertImageFromCanvas();
      });
    } else {
      insertImageFromCanvas();
    }
  };

  const getCanvasURL = () => {
    const dataValue = ref?.current?.['children']?.[0];
    if (dataValue) {
      let str = (dataValue as HTMLCanvasElement).toDataURL();
      str = str.split('data:image/png;base64,')[1];
      return str;
    }
  };

  return (
    <>
      <InputWrapper updateText={setText} value={text} label="QR code text" />
      <Slider
        className={styles.slider}
        label="Size"
        onChange={(v: number) => setSize(v)}
        min={50}
        max={300}
        step={5}
        defaultValue={125}
        showValue
        snapToStep
      />
      <div ref={ref} data-testid="qrcode-canvas">
        <QRCode size={size} value={text} />
      </div>
      <div className={styles.button}>
        <Button appearance="primary" onClick={insertImage}>
          Insert Image
        </Button>
      </div>
    </>
  );
}
