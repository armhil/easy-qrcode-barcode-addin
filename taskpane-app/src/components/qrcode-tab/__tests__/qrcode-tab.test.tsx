import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { AddinUtils, LoggingUtils } from 'easy-addins-utils';
import { QrCodeTab } from '../qrcode-tab';

describe('qrcode rendering', () => {
  it('should match with snapshot', () => {
    const tree = renderer.create(<QrCodeTab />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should log a trace', () => {
    const traceMockFn = jest.fn();
    LoggingUtils.Trace = traceMockFn;

    render(<QrCodeTab />);
    expect(traceMockFn).toHaveBeenCalledTimes(1);
    expect(traceMockFn).toHaveBeenCalledWith('qrbar-qrcode');
  });

  it('should render links with correct attributes', () => {
    const dom = render(<QrCodeTab />);

    const links = dom.queryAllByRole('link');
    // no links on the page
    expect(links.length).toBe(0);
  });

  it('should render the canvas correctly', () => {
    const dom = render(<QrCodeTab />);

    const canvas = dom.getByTestId('qrcode-canvas');
    expect(canvas).not.toBeEmptyDOMElement();
    expect((canvas.firstChild as HTMLElement).tagName).toBe('CANVAS');
  });

  it('should attempt to read from document if textbox is empty', () => {
    AddinUtils.InsertImage = jest.fn();
    AddinUtils.GetText = jest.fn();

    const dom = render(<QrCodeTab />);

    act(() => {
      const insertButton = dom.queryByRole('button');
      insertButton!.click();
      expect(AddinUtils.GetText).toHaveBeenCalledTimes(1);
    });
  });

  it('should attempt to insert image if textbox is not empty', () => {
    AddinUtils.InsertImage = jest.fn();
    AddinUtils.GetText = jest.fn();
    const dom = render(<QrCodeTab />);

    act(() => {
      const input = dom.queryAllByRole('textbox');

      fireEvent.change(input[0] as HTMLElement, { target: { value: 'testing' } });
      const insertButton = dom.queryByRole('button');
      insertButton!.click();
      expect(AddinUtils.GetText).toHaveBeenCalledTimes(0);
      expect(AddinUtils.InsertImage).toHaveBeenCalledTimes(1);
    });
  });
});
