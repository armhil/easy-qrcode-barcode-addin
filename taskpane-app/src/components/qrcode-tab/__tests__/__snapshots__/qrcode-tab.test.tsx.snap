// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`qrcode rendering should match with snapshot 1`] = `
[
  <label
    className="fui-Label ___1yevxz2_yljzsa0 fk6fouc f19n0e5 fkhj508 f1i3iumi"
    htmlFor="fui-r0"
  >
    <b>
      QR code text
    </b>
  </label>,
  <span
    className="fui-Input r1jtohuq ___gxxrxg0_glgvo90 fvcxoqz f1ub3y4t f1l4zc64 f1m52nbi f8vnjqi fz1etlk f1klwx88 f1hc16gm fly5x3f ffemd8d"
  >
    <input
      className="fui-Input__input rvp2gzh"
      id="fui-r0"
      onChange={[Function]}
      placeholder="Type here or select from your document"
      role="textbox"
      type="text"
      value=""
    />
  </span>,
  <div
    className="ms-Slider ms-Slider-enabled ms-Slider-row ___8jbu6p0_0000000 f18og1rn f1dl2k97 root-109"
  >
    <label
      className="ms-Label titleLabel-120"
      htmlFor="Slider0"
    >
      Size
    </label>
    <div
      className="ms-Slider-container container-111"
    >
      <div
        aria-disabled={false}
        aria-label="Size"
        aria-valuemax={300}
        aria-valuemin={50}
        aria-valuenow={125}
        aria-valuetext="125"
        className="ms-Slider-slideBox ms-Slider-showValue slideBox-112"
        data-is-focusable={true}
        id="Slider0"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        role="slider"
        tabIndex={0}
      >
        <div
          className="ms-Slider-line line-114"
        >
          <span
            className="ms-Slider-thumb thumb-113"
            style={
              {
                "left": "30%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-inactive inactiveSection-117"
            style={
              {
                "width": "0%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-active activeSection-116"
            style={
              {
                "width": "30%",
              }
            }
          />
          <span
            className="lineContainer-115 ms-Slider-inactive inactiveSection-117"
            style={
              {
                "width": "70%",
              }
            }
          />
        </div>
      </div>
      <label
        className="ms-Label ms-Slider-value valueLabel-121"
      >
        125
      </label>
    </div>
  </div>,
  <div
    data-testid="qrcode-canvas"
  >
    <canvas
      height={125}
      style={
        {
          "height": 125,
          "width": 125,
        }
      }
      width={125}
    />
  </div>,
  <div
    className="___15qz5iz_0000000 f18og1rn"
  >
    <button
      className="fui-Button r1alrhcs ___1akj6hk_ih97uj0 ffp7eso f1p3nwhy f11589ue f1q5o8ev f1pdflbu f1phragk f15wkkf3 f1s2uweq fr80ssc f1ukrpxl fecsdlb f1rq72xc fnp9lpt f1h0usnq fs4ktlq f16h9ulv fx2bmrt f1d6v5y2 f1rirnrt f1uu00uk fkvaka8 f1ux7til f9a0qzu f1lkg8j3 fkc42ay fq7113v ff1wgvm fiob0tu f1j6scgf f1x4h75k f4xjyn1 fbgcvur f1ks1yx8 f1o6qegi fcnxywj fmxjhhp f9ddjv3 f17t0x8g f194v5ow f1qgg65p fk7jm04 fhgccpy f32wu9k fu5nqqq f13prjl2 f1czftr5 f1nl83rv f12k37oa fr96u23"
      onClick={[Function]}
      type="button"
    >
      Insert Image
    </button>
  </div>,
]
`;
